core/include/cubos/core/ecs/dispatcher.hpp

```diff
@@ -9,81 +9,198 @@
 #include <string>
 #include <memory>
 
+#define ENSURE_CURR_SYSTEM()                                                                                           \
+    if (!currSystem)                                                                                                   \
+    {                                                                                                                  \
+        CUBOS_ERROR("No system currently selected!");                                                                  \
+        return;                                                                                                        \
+    }
+#define ENSURE_CURR_TAG()                                                                                              \
+    if (!currTagSettings)                                                                                              \
+    {                                                                                                                  \
+        CUBOS_ERROR("No tag currently selected!");                                                                     \
+        return;                                                                                                        \
+    }
+#define ENSURE_SYSTEM_SETTINGS(obj)                                                                                    \
+    if (!obj->settings)                                                                                                \
+    {                                                                                                                  \
+        obj->settings = std::make_shared<SystemSettings>();                                                            \
+    }
+
 namespace cubos::core::ecs
 {
     /// @brief Dispatcher is a class that is used to register systems and call them all at once, where they will be
     /// executed in order of the stages they are in.
     class Dispatcher
     {
     public:
-        // Direction enum
-        enum class Direction
+        struct Dependency;
+        struct SystemSettings;
+        struct System;
+
+        /// Internal class to specify system dependencies
+        struct Dependency
         {
-            Before,
-            After
+            std::vector<std::string> tag;
+            std::vector<System*> system;
```

Comment:
I think maybe its safer to refer to systems by their index on the vector instead of by a pointer.

Reply:
This isn't reliable as systems are moved from the `pendingSystems` to `systems`. Also this dependency class effectively "disappears" once the call chain is built; the dependency order is implicit with the order systems are in the `systems` array, so these values aren't used.

I did forget about this fact and didn't delete them though, so I'll change that.